File: frontend/src/types/simulation.ts
export interface HealthResponse {
  status: string;
  model_loaded: boolean;
  device: string;
  version: string;
}

export interface SeedsResponse {
  seeds: number[];
}

export interface SimulationRequest {
  seed: number;
  mode: 'live' | 'replay';
}

export interface MetricsSummary {
  final_survival: number;
  population_saved: number;
  initial_population: number;
  total_reward: number;
  agent_reliability: number;
  active_emergencies: number;
  resource_efficiency: number;
}

export interface EventItem {
  step: number;
  text: string;
  type: 'info' | 'warning' | 'critical' | 'success';
}

export interface DriftStatus {
  status: 'NORMAL' | 'WARNING' | 'RECOVERING' | 'STABLE';
  api_status: 'active' | 'deprecated';
  current_schema_version: number;
  last_error: string | null;
}

export interface StepData {
  step: number;
  grid: number[][][]; // 5x5x4 [population, severity, resources, zone_id]
  cmd_msg: {
    intent: string;
    zone: number;
    resource: string;
    priority: string;
    units: number;
  } | null;
  res_action: {
    action: string;
    zone: number;
    units: number;
    resource: string;
  } | null;
  reward: number;
  total_reward: number;
  survival_rate: number;
  mean_severity: number;
  max_severity: number;
  drift_status: DriftStatus;
  oversight_logs: any[];
  decision_explanation: string;
}

export interface SimulationResponse {
  agent_type: string;
  seed: number;
  mode: 'live' | 'replay';
  steps: StepData[];
  survival_curve: number[];
  severity_curve: number[];
  events: EventItem[];
  metrics: MetricsSummary;
}

export interface ComparisonMeta {
  survival_delta: number;
  population_saved_delta: number;
  policies_match: boolean;
}

export interface ComparisonResponse {
  seed: number;
  mode: string;
  trained: SimulationResponse;
  random: SimulationResponse;
  comparison: ComparisonMeta;
}

export interface WSFrame {
  type: 'init' | 'step' | 'complete' | 'error';
  step?: number;
  data?: any;
  seed?: number;
  total_steps?: number;
  metrics?: MetricsSummary;
  events?: EventItem[];
  message?: string;
}

File: backend/models/schemas/simulation.py
"""
backend/models/schemas/simulation.py
Pydantic schemas for API requests, responses, grid steps, metrics, and WebSocket frames.
"""

from typing import List, Dict, Any, Optional
from pydantic import BaseModel, Field


class HealthResponse(BaseModel):
    status: str = "ok"
    model_loaded: bool = False
    device: str = "cpu"
    version: str = "2.0.0"


class SeedsResponse(BaseModel):
    seeds: List[int] = Field(default_factory=lambda: [123, 42, 999])


class SimulationRequest(BaseModel):
    seed: int = Field(default=123, description="RNG seed for scenario reproducibility")
    mode: str = Field(default="live", description="Execution mode: 'live' or 'replay'")


class MetricsSummary(BaseModel):
    final_survival: float
    population_saved: int
    initial_population: int
    total_reward: float
    agent_reliability: float
    active_emergencies: int
    resource_efficiency: float


class EventItem(BaseModel):
    step: int
    text: str
    type: str  # 'info' | 'warning' | 'critical' | 'success'


class SimulationResponse(BaseModel):
    agent_type: str = "trained"
    seed: int
    mode: str
    steps: List[Dict[str, Any]]
    survival_curve: List[float]
    severity_curve: List[float]
    events: List[EventItem]
    metrics: MetricsSummary


class ComparisonMeta(BaseModel):
    survival_delta: float = Field(..., description="Trained final survival minus random final survival (percentage points)")
    population_saved_delta: int = Field(..., description="Trained survivors saved minus random survivors saved")
    policies_match: bool = Field(..., description="True if trained and random trajectories are identical (should be False)")
    decision_similarity: float = Field(..., description="Fraction of timesteps where trained and random actions matched (0.0 to 1.0)")


class ComparisonResponse(BaseModel):
    """Response containing both trained and random baseline trajectories for side-by-side comparison."""
    seed: int
    mode: str
    trained: SimulationResponse
    random: SimulationResponse
    comparison: ComparisonMeta


class WSFrame(BaseModel):
    type: str  # "init" | "step" | "complete" | "error"
    step: Optional[int] = None
    data: Dict[str, Any]
